pkg/app/pipedv1/livestatereporter/livestatereporter.go

```diff
@@ -0,0 +1,171 @@
+// Copyright 2024 The PipeCD Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package livestatereporter provides a piped component
+// that reports the changes as well as full snapshot about live state of registered applications.
+package livestatereporter
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"go.uber.org/zap"
+	"golang.org/x/sync/errgroup"
+	"google.golang.org/grpc"
+
+	"github.com/pipe-cd/pipecd/pkg/app/server/service/pipedservice"
+	"github.com/pipe-cd/pipecd/pkg/model"
+	pluginapi "github.com/pipe-cd/pipecd/pkg/plugin/api/v1alpha1"
+	"github.com/pipe-cd/pipecd/pkg/plugin/api/v1alpha1/livestate"
+)
+
+type applicationLister interface {
+	ListByPluginName(name string) []*model.Application
+}
+
+type apiClient interface {
+	ReportApplicationLiveState(ctx context.Context, req *pipedservice.ReportApplicationLiveStateRequest, opts ...grpc.CallOption) (*pipedservice.ReportApplicationLiveStateResponse, error)
+	ReportApplicationSyncState(ctx context.Context, req *pipedservice.ReportApplicationSyncStateRequest, opts ...grpc.CallOption) (*pipedservice.ReportApplicationSyncStateResponse, error)
+}
+
+// Reporter represents a component that reports the snapshot about live state of registered applications.
+type Reporter interface {
+	Run(ctx context.Context) error
+}
+
+type reporter struct {
+	reporters []pluginReporter
+	logger    *zap.Logger
+}
+
+// NewReporter creates a new reporter.
+func NewReporter(appLister applicationLister, apiClient apiClient, plugins map[string]pluginapi.PluginClient, logger *zap.Logger) Reporter {
+	rlogger := logger.Named("live-state-reporter")
+	r := &reporter{
+		reporters: make([]pluginReporter, 0, len(plugins)),
+		logger:    rlogger,
+	}
+
+	for name, p := range plugins {
+		r.reporters = append(r.reporters, pluginReporter{
+			pluginName:            name,
+			snapshotFlushInterval: time.Minute,
+			appLister:             appLister,
+			apiClient:             apiClient,
+			pluginClient:          p,
+			logger:                rlogger.With(zap.String("plugin-name", name)),
+		})
+	}
+	return r
+}
+
+func (r *reporter) Run(ctx context.Context) error {
```

Comment:
Implemented it based on this code.
https://github.com/pipe-cd/pipecd/blob/336d96d501311103dd9921bcea55722e2c8e774a/pkg/app/piped/livestatereporter/reporter.go#L104C1-L127C2